Automattic/wp-calypso
# Support dedicated /me/security UI to change email address

#### Changes proposed in this Pull Request

This PR may well be too big, but it takes on a number of closely related tasks to expand the work in #43527:
* It updates the state and data layers for user settings as follows:
  - Settings updates now support an error handler callback
  - There is a new `isFetchingUserSettings()` selector (and the related wiring) so fetches and updates can be reflected in the UI
  - The `isPendingEmailChange()` selector has been updated to force a boolean return value.
* It refactors the email logic currently used in `/me/account` into a dedicated component
  - At the same time, it does _not_refactor the existing code to use Redux, so it has plenty of bifurcation depending on whether the container is supplying a `UserSettings` object or expecting the component to rely on the Redux state.
* It adds a new component to `/me/security/account-email` that is only responsible for updating the account email address, and fits the UX for the other menu items listed under `/me/security`.

#### Testing instructions

There is a LOT to test here.

##### Regression Tests for `/me/account`

Our main goal here is to make sure that the existing email modification UI continues to work. Make sure the following work:
* Navigate to `/me/account` and update the email address field.
  - Make sure that invalid addresses result in an error being shown, and the submit button being disabled.
  - Make sure that if the email address is the same as your starting email address, the submit button is disabled.
  - Enter a valid email address and submit the updated email address.
* Ensure that after you've submitted the change, a pending change warning appears and you can't modify the field directly.
* Click on the "Cancel" option for the pending change, and verify that the form updates and reverts to the original email address, and you can update and submit the email address again.
It may also be worth verifying an email to check that the UI updates appropriately -- this is existing behaviour, but I don't believe I touched it in this case.

##### Testing the email option in `/me/security`

Using a local Calypso version, or by using the [calypso.live](https://calypso.live/?branch=try/security-account-email-ui) UI for this branch, try the following:
* Navigate to `/me/security`.
* Verify that clicking on the "Account Email" entry takes you to `/me/security/account-email`, and that clicking the "Back" button gets you back to the menu. Click on the "Account Email" entry again.
* In the dedicated email UI, verify that the following occur:
  - When the email address in the text box matches your initial email address, the "Update" button is disabled.
  - When the email address in the text box is invalid, an appropriate error should be displayed, and the "Update" button should be disabled.
  - When you type in a valid email address, the "Update" button should be enabled.
* Type in a valid email address for yourself, and click on the "Update" button.
* The button text should change to "Updating..." and should be disabled while that text displays. Once the update completes, the button should remain disabled with the original "Update" text. In addition, the email field should be disabled and should show a prominent notice indicating that the email change is still pending.
* Click on the "Back" button.
* Verify that the "Account Email" entry shows a warning icon and the text: "There is an unconfirmed change of your email address to **email@example.com**."
* Click on the "Account Email" entry, and verify that the pending email change is displayed.
* Click on the "Cancel" text/link in the email warning, and verify that the email field reverts to your original email address.
* Click on the "Back" button, and verify that the main security menu shows a check-mark and your original email address.
